_posts/gRPC-Go performance improvements.md

```diff
@@ -0,0 +1,148 @@
+
+
+### Expanding stream window on receiving large messages 
+
+[Code link](https://github.com/grpc/grpc-go/pull/1248)
+
+This is an optimization used by gRPC-C to achieve performance benefits for large messages. The idea is that when there's an active read by the application on the receive side, we can effectively bypass stream-level flow control to request the whole message. This proves to be very helpful when the message is big. Since the application is already committed to reading and has allocated enough memory for it, it makes sense that we send a proactive large window update (if necessary) to get the whole message rather than receiving it in chunks and sending window updates when we run low on window.
+
+This optimization alone provided a 10x improvement for large messages on high-latency networks.
+
+
+### Decoupling application reads from connection flow control
+
+[Code link](https://github.com/grpc/grpc-go/pull/1265)
+
+After having several discussions with gRPC-Java and gRPC-C core team we realized that gRPC-Go's connection-level flow control was overly restrictive in the sense that window updates on the connection depended upon if the application had read data from it or not. It must be noted that it makes perfect sense to have stream-level flow control depended on application read but not so much for connection-level flow control. The rationale is as floows: A connection is shared by several streams (RPCs). Now, if there were at least one stream that read slowly or didn't read at all, it would hamper the performance or completely stall other streams on that connection. This happens because we won't send out window updates on the connection until that slow or inactive stream read data. Therefore, it makes sense to decouple the connection's flow control from application reads. 
+
+However, this begs at least two questions: 1. Won't a client be able to send as much data as it wants to the server by creating new streams when one runs out? 2. Why even have connection-level flow control if the stream-level flow control is enough.
+
+The answer to the first question is short and simple: no. A server has an option to limit the number of streams that it intends to server concurrently. Therefore, although at first it may seem like a problem, it really is not.
+
+The second question is very interesting: why even bother with connection-level flow control?
+
+It is true that stream-level flow control is sufficient to throttle a sender from sending too much data. But not having connection-level flow control (or using an unlimited connection-level window) makes it so that when things get slower on a stream, opening a new one will appear to make things faster. This is a slippery slope and will only take one so far. However, having a connection-level flow control window set to the Bandwidth Delay Product (BDP) of the network puts an upper-bound on how much performance can realistically be squeezed out of the network.
```

Comment:
Why is it a slippery slope.. its not clearly expressed here. So I would suggest replacing that line with a brief mention of the major problem this causes. Eg:
If the rationale is that we are bounded by a number of streams per connection and that  things artificially get faster, because the intention is to actually flow control at the connection level, leading to cascading flow control on streams..